lib/storage.js

```diff
@@ -0,0 +1,111 @@
+'use strict';
+
+const UPDATE_TIMEOUT_MS = 500;
+
+/**
+ * Check to if the given value is mergable.
+ *
+ * @param {*} value
+ *
+ * @return {Boolean}  Returns true if the object can be merged, false otherwise.
+ */
+function canMerge(value) {
+  return value !== null && typeof value === 'object' && !Array.isArray(value);
+}
+
+/**
+ * Check if the given value will overwrite other values during a merge.
+ *
+ * @param {*} value
+ *
+ * @return {Boolean} Returns true if the value will overwrite others, false otherwise.
+ */
+function canOverwriteWith(value) {
+  return value !== null && value !== undefined && !(value instanceof Function);
+}
+
+/**
+ * Perform a deep merge of objects based on key.
+ * Warning! Destination is modified in place.
+ *
+ * @param {Object} source       Object from which keys are copied out of
+ * @param {Object} destination  Object from which keys are copied into
+ */
+function merge(source, destination) {
+  if (canMerge(source)) {
+    for (const key in source) {
+      if (source.hasOwnProperty(key)) {
+        const sourceValue = source[key];
+        const destinationValue = destination[key];
+
+        if (canOverwriteWith(sourceValue)) {
+          if (canMerge(sourceValue) && canMerge(destinationValue)) {
+            merge(sourceValue, destinationValue);
+          } else {
+            destination[key] = sourceValue; // eslint-disable-line no-param-reassign
+          }
+        }
+      }
+    }
+  } else if (canOverwriteWith(source)) {
+    destination = source; // eslint-disable-line no-param-reassign
+  }
+}
+
+class Storage {
+  /**
+   * The storage engine maintains a merged set of properties retrived from
+   * source plugins. An optional EventEmitter can be provided. When properties
+   * change an "update" event will fire on the emitter with the new properites.
+   *
+   * @param {EventEmitter} emitter
+   */
+  constructor(emitter) {
```

Comment:
Why accept an emitter as an arg instead of just extending `EventEmitter`? The `Storage` object seems like it will stick around for the lifetime of the process, so why not have other things subscribe to it rather than vice-versa?